### PR TITLE
WebXR Plane Detection

### PR DESCRIPTION
**Pull request by Maksims**
Implementation of an experimental draft spec: https://immersive-web.github.io/real-world-geometry/plane-detection.html

It is available in Chrome for Android 89, with enabled chrome://flags#webxr-incubations flag.

Plane detection provides information about real-world surfaces estimations in form of planes.
Each plane can have its position, rotation and a list of edge points making a complex polygon.

### New APIs:
```js
// pc.XrManager
app.xr.planeDetection // interface to access plane detection
app.xr.start(camera, ..., ..., {
    planeDetection: true
}); // new option to enable plane detection

// pc.XrPlaneDetection
planeDetection.supported // true if plane detection is supported
planeDetection.available // true if plane detection is available
planeDetection.planes // a list of pc.XrPlane's

planeDetection.on('available', () => { }); // fired when plane detection becomes available, due to spec design it is not possible to know if it is available before a first XR session update
planeDetection.on('unavailable', () => { }); // fired when plane detection becomes unavailable, for example when session is ended
planeDetection.on('add', (plane) => { }); // fired when a new plane is added
planeDetection.on('remove', (plane) => { }); // fired when a plane is removed

// pc.XrPlane
plane.id // numerical id of a plane
plane.orientation // 'horizontal' or 'vertical' orientation of a plane, or null if any other orientation
plane.points // a list of edge points of a plane. Their coords are in local space to a plane
plane.getPosition() // get world position of a plane
plane.getRotation() // get world rotation of a plane

plane.once('remove', () => { }); // fired when plane is removed
plane.on('change', () => { }); // fired when plane orientation or points have been changed
```

### Test project:
https://playcanvas.com/project/782753/overview/webxr-plane-detection
This project builds mesh, updates it when the plane is changed and draw lines for each detected plane.

### Test build:
https://playcanv.as/p/f2ESRGge/
Ensure you have chrome://flags#webxr-incubations enabled

### Video:
https://twitter.com/mrmaxm/status/1379802592163864581

thumb

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

**Comment by Maksims** [comment]
I'm not sure how to deal with a TypeScript error as it fails to find XRPlane, and not sure where to add a "polyfill" for such a test. Please advice @mvaligursky

File: src/xr/xr-plane.js [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+let ids = 0;
+
+/**
+ * @class
+ * @name XrPlane
+ * @classdesc Detected Plane instance that provides position, rotation and polygon points. Plane is a subject to change during its lifetime.
+ * @description Detected Plane instance that provides position, rotation and polygon points. Plane is a subject to change during its lifetime.
+ * @param {XrPlaneDetection} planeDetection - Plane detection system.
+ * @param {XRPlane} xrPlane - XRPlane that is instantiated by WebXR system.
```
Comment (mvaligursky) [inline src/xr/xr-plane.js:13]: should type "XRPlane" perhaps be changed to {object} to fix the type issue?

File: src/xr/xr-plane-detection.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+     *     // new plane is removed
+     * });
+     */
+
+    _onSessionEnd() {
+        for (let i = 0; i < this._planes.length; i++) {
+            this._planes[i].destroy();
+        }
+        this._planesIndex.clear();
+        this._planes = [];
```
Comment (mvaligursky) [inline src/xr/xr-plane-detection.js:78]: maybe just set to null instead of allocating new array

Comment (mvaligursky) [reply src/xr/xr-plane-detection.js:78]: actually .. this is not destroy() .. so maybe just `this._planes.length = 0;`

File: src/xr/xr-plane-detection.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+ */
+class XrPlaneDetection extends EventHandler {
+    constructor(manager) {
+        super();
+
+        this._manager = manager;
+        this._supported = !! window.XRPlane;
+        this._available = false;
+
+        this._planesIndex = new Map();
```
Comment (mvaligursky) [inline src/xr/xr-plane-detection.js:31]: it'd be great to have comment for the Map here, with info about what is the key and what is the value

File: src/xr/xr-plane-detection.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                this._available = true;
+                this.fire('available');
+            } catch (ex) {
+                return;
+            }
+        } else {
+            detectedPlanes = frame.detectedPlanes;
+        }
+
+        for (const [xrPlane, plane] of this._planesIndex) {
```
Comment (mvaligursky) [inline src/xr/xr-plane-detection.js:101]: add comment on what the loop does please

File: src/xr/xr-plane-detection.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            if (! detectedPlanes.has(xrPlane)) {
+                this._planesIndex.delete(xrPlane);
+                this._planes.splice(this._planes.indexOf(plane), 1);
+                plane.destroy();
+                this.fire('remove', plane);
+            }
+        }
+
+        for (const xrPlane of detectedPlanes) {
```
Comment (mvaligursky) [inline src/xr/xr-plane-detection.js:110]: add comment on what the loop does please

File: src/xr/xr-plane.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+    /**
+     * @function
+     * @name XrPlane#getRotation
+     * @description Get the world space rotation of a plane.
+     * @returns {Vec3} The world space rotation of a plane.
+     */
+    getRotation() {
+        return this._rotation;
+    }
```
Comment (mvaligursky) [inline src/xr/xr-plane.js:92]: I'm not sure about the contention in the Xr codebase, but should these be property getters instead of functions?

Comment (Maksims) [reply src/xr/xr-plane.js:92]: These are functions to be consistent with other XR modules as well as PlayCanvas codebase, where they are functions.

**Review by mvaligursky — APPROVED** [review]

**Comment by yaustar** [comment]
Self note: add project https://playcanvas.com/project/782753/overview/webxr-plane-detection to tutorials section

File: src/xr/xr-plane-detection.js [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+     * @example
+     * app.xr.planeDetection.on('add', function (plane) {
+     *     // new plane is added
+     * });
+     */
+
+    /**
+     * @event
+     * @name XrPlaneDetection#remove
+     * @description Fired when a {@link XrPlane} is removed to the list.
```
Comment (willeastcott) [inline src/xr/xr-plane-detection.js:68]:
```suggestion
     * @description Fired when a {@link XrPlane} is removed from the list.
```

File: src/xr/xr-plane-detection.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
+import { EventHandler } from '../core/event-handler.js';
+import { XrPlane } from './xr-plane.js';
+
+/**
+ * @class
+ * @name XrPlaneDetection
+ * @classdesc Plane Detection provides the ability to detect real world surfaces based on estimations of underlying AR system.
+ * @description Plane Detection provides the ability to detect real world surfaces based on estimations of underlying AR system.
```
Comment (willeastcott) [inline src/xr/xr-plane-detection.js:8]:
```suggestion
 * @classdesc Plane Detection provides the ability to detect real world surfaces based on estimations of the underlying AR system.
 * @description Plane Detection provides the ability to detect real world surfaces based on estimations of the underlying AR system.
```

File: src/xr/xr-plane-detection.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+ * @param {XrManager} manager - WebXR Manager.
+ * @property {boolean} supported True if Plane Detection is supported.
+ * @property {boolean} available True if Plane Detection is available. This property can be set to true only during running session.
```
Comment (willeastcott) [inline src/xr/xr-plane-detection.js:11]:
```suggestion
 * @property {boolean} available True if Plane Detection is available. This property can be set to true only during a running session.
```

File: src/xr/xr-plane-detection.js [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+ * @property {XrPlane[]|null} planes List of {@link XrPlane} that contain individual plane information, or null if plane detection isn't available.
```
Comment (willeastcott) [inline src/xr/xr-plane-detection.js:12]:
```suggestion
 * @property {XrPlane[]|null} planes Array of {@link XrPlane} instances that contain individual plane information, or null if plane detection is not available.
```

File: src/xr/xr-plane.js [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+/**
+ * @class
+ * @name XrPlane
+ * @classdesc Detected Plane instance that provides position, rotation and polygon points. Plane is a subject to change during its lifetime.
+ * @description Detected Plane instance that provides position, rotation and polygon points. Plane is a subject to change during its lifetime.
+ * @param {XrPlaneDetection} planeDetection - Plane detection system.
+ * @param {object} xrPlane - XRPlane that is instantiated by WebXR system.
+ * @property {number} id Unique identifier of a plane.
+ * @property {string|null} orientation Plane's pecific orientation (horizontal or vertical) or null if orientation is anything else.
```
Comment (willeastcott) [inline src/xr/xr-plane.js:15]:
```suggestion
 * @property {string|null} orientation Plane's specific orientation (horizontal or vertical) or null if orientation is anything else.
```

File: src/xr/xr-plane.js [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+     * @example
+     * plane.once('remove', function () {
+     *     // plane is not available anymore
+     * });
+     */
+
+    /**
+     * @event
+     * @name XrPlane#change
+     * @description Fired when {@link XrPlane} attributes such as: orientation and/or points have been changed. Position and Rotation can change at any time without triggering `change` event.
```
Comment (willeastcott) [inline src/xr/xr-plane.js:47]:
```suggestion
     * @description Fired when {@link XrPlane} attributes such as: orientation and/or points have been changed. Position and rotation can change at any time without triggering a `change` event.
```

File: src/xr/xr-plane.js [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+     */
+    getPosition() {
+        return this._position;
+    }
+
+    /**
+     * @function
+     * @name XrPlane#getRotation
+     * @description Get the world space rotation of a plane.
+     * @returns {Vec3} The world space rotation of a plane.
```
Comment (willeastcott) [inline src/xr/xr-plane.js:88]:
```suggestion
    * @returns {Quat} The world space rotation of a plane.
```

File: src/xr/xr-plane.js [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    get orientation() {
+        return this._orientation;
+    }
+
+    /**
+     * @name XrPlane#points
+     * @type {object[]}
+     * @description List of DOMPointReadOnly objects that is an object with `x y z` properties that define local point of a planes polygon.
```
Comment (willeastcott) [inline src/xr/xr-plane.js:105]:
```suggestion
     * @description List of DOMPointReadOnly objects. DOMPointReadOnly is an object with `x y z` properties that defines a local point of a plane's polygon.
```

**Review by willeastcott — APPROVED** [review]

**Comment by willeastcott** [comment]
OK, LGTM now, @Maksims. Can you resolve the `xr-manager.js` conflict and I'll merge.

**Comment by willeastcott** [comment]
Looks like Typescript definitions are failing:

image

**Comment by Maksims** [comment]
> Looks like Typescript definitions are failing:
> 
> image

I have no idea how to fix those TypeScript issues.

File: src/xr/xr-manager.js [hunk 15]
```diff
@@ -181,6 +183,9 @@ class XrManager extends EventHandler {
      * });
      * @param {object} [options] - Object with additional options for XR session initialization.
      * @param {string[]} [options.optionalFeatures] - Optional features for XRSession start. It is used for getting access to additional WebXR spec extensions.
+     * @param {boolean} [options.depthSensing] - Set to true to attempt to enable {@link XrDepthSensing}.
```
Comment (willeastcott) [inline src/xr/xr-manager.js:186]: This is a duplicate of the param definition at line 190

Comment (willeastcott) [reply src/xr/xr-manager.js:186]: Except that declaration is an object

**Pull request merged**